pyro/infer/mcmc/mcmc.py

```diff
@@ -278,8 +251,61 @@ def __init__(self, kernel, num_samples, warmup_steps=0,
                                             num_chains, mp_context)
         else:
             self.sampler = _SingleSampler(kernel, num_samples, warmup_steps)
-        super(MCMC, self).__init__()
 
     def _traces(self, *args, **kwargs):
         for sample in self.sampler._traces(*args, **kwargs):
             yield sample
+
+    def marginal(self, sites=None):
+        return _EmpiricalMarginal(self, sites)
+
+
+class _EmpiricalMarginal(object):
```

Comment:
I think we can make this inherited from `abstract_infer.EmpiricalMarginal`.

Reply:
`EmpiricalMarginal` inherits from the `Empirical` distribution, so that won't work (at least not without creating other problems). The main issue is that this needs to hold an `Empirical` distribution per latent, therefore it cannot inherit from `Empirical` distribution without relaxing the shape requirements for latents.

Reply:
Also, if we make `EmpiricalMarginal` not inherit from Empirical, it creates issues with importance sampling.